File: exercism/typescript/phone-number/phone-number.ts
export function clean(dirtyNumber: string): string {
  let cleanedNumber = dirtyNumber.replaceAll(new RegExp("[-\.\ \(\)\+]", "g"), "");

  if (cleanedNumber.match(new RegExp("[a-zA-Z]", "g"))) {
    throw new Error('Letters not permitted');
  }
  if (cleanedNumber.match(new RegExp("[@:!]", "g"))) {
    throw new Error('Punctuations not permitted');
  }


  if (cleanedNumber.length < 10) {
    throw new Error('Incorrect number of digits');
  }

  if (cleanedNumber.length === 11) {
    if (cleanedNumber[0] === "1") {
      cleanedNumber = cleanedNumber.substring(1);
    } else {
      throw new Error('11 digits must start with 1');
    }
  }

  if (cleanedNumber.length === 10) {
    if (cleanedNumber[0] === "0") {
      throw new Error('Area code cannot start with zero');
    } else if (cleanedNumber[0] === "1") {
      throw new Error('Area code cannot start with one');
    }
    if (cleanedNumber[3] === "0") {
      throw new Error('Exchange code cannot start with zero');
    } else if (cleanedNumber[3] === "1") {
      throw new Error('Exchange code cannot start with one');
    }
    return cleanedNumber;
  } else {
    throw new Error('More than 11 digits');
  }
}

File: exercism/python/phone-number/phone_number.py
class PhoneNumber:
    def __init__(self, number: str):
        number = self.sanitize_number(number)
        number = self.validate_phone_number_phase1(number)
        self.number = number
        self.area_code: str = number[:3]
        self.exchange_code: str = number[3:6]
        self.subscriber_number: str = number[6:]
        self.validate_phone_number_codes()

    @staticmethod
    def sanitize_number(number: str) -> str:
        translation_table = str.maketrans("", "", " .+-()")
        number = number.translate(translation_table)
        return number

    @staticmethod
    def validate_phone_number_phase1(number: str) -> str:
        min_length = 10
        max_length = 11
        icc = "1"  # international country code
        if len(number) < min_length:
            raise ValueError(f"must not be fewer than {min_length} digits")
        if len(number) > max_length:
            raise ValueError(f"must not be greater than {max_length} digits")
        if len(number) == max_length:
            if not number.startswith(icc):
                raise ValueError(f"{max_length} digits must start with {icc}")
            else:
                number = number[1:]
        if not number.isdigit():
            if any(char in ["@", ":", "!", "?", ","] for char in number):
                raise ValueError("punctuations not permitted")
            else:
                raise ValueError("letters not permitted")
        return number

    def validate_phone_number_codes(self) -> None:
        if self.exchange_code.startswith("0"):
            raise ValueError("exchange code cannot start with zero")
        if self.exchange_code.startswith("1"):
            raise ValueError("exchange code cannot start with one")
        if self.area_code.startswith("0"):
            raise ValueError("area code cannot start with zero")
        if self.area_code.startswith("1"):
            raise ValueError("area code cannot start with one")

    def pretty(self) -> str:
        return f"({self.area_code})-{self.exchange_code}-{self.subscriber_number}"
